### PR TITLE
Remove dependency on ICompilationOptionsProvider

### PR DESCRIPTION
**Pull request by davidfowl**
- Removed Configuration from RazorViewEngineOptions
- Added ParseOptions and CompilationOptions to RazorViewEngineOptions

#3633

/cc @pranavkm @pakrym 

There's a broken test that shows flowing compilation options from the project file. I can update the functional test project to support passing setting the right preprocessor symbols at runtime.

EDIT:

Update the test

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs [hunk 1]
```diff
@@ -15,7 +16,9 @@ public class RazorViewEngineOptions
     {
         private IFileProvider _fileProvider;
 
-        private string _configuration;
+        private CSharpParseOptions _parseOptions = new CSharpParseOptions(LanguageVersion.CSharp6);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: `readonly`

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: nevermnd. It's settable.

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs [hunk 2]
```diff
@@ -49,42 +52,54 @@ public IFileProvider FileProvider
...
         }
 
         /// <summary>
-        /// Gets or sets the callback that is used to customize Razor compilation
-        /// to change compilation settings you can update <see cref="RoslynCompilationContext.Compilation"/> property.
-        /// Customizations made here would not reflect in tooling (Intellisense).
+        /// Gets or sets the <see cref="CSharpParseOptions"/> options used by Razor view compilation.
         /// </summary>
-        public Action<RoslynCompilationContext> CompilationCallback
+        public CSharpParseOptions ParseOptions
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: Granted we can make this work, but is there benefit in letting users customize this? Razor parser had to be modified to switch to C# 6, so changing the language version here might not be sufficient.

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs [hunk 3]
```diff
@@ -258,13 +254,16 @@ private List<MetadataReference> GetApplicationReferences()
...
-                // Microsoft.Extensions.Runtime.Roslyn/RoslynCompiler.cs#L164
-                // We don't want to take a dependency on the Roslyn bit directly since it pulls in more dependencies
-                // than the view engine needs (Microsoft.Extensions.Runtime) for example
-                references.Add(ConvertMetadataReference(metadataReference));
+                foreach (var metadataReference in export.MetadataReferences)
+                {
+                    // Taken from https://github.com/aspnet/KRuntime/blob/757ba9bfdf80bd6277e715d6375969a7f44370ee/src/...
+                    // Microsoft.Extensions.Runtime.Roslyn/RoslynCompiler.cs#L164
+                    // We don't want to take a dependency on the Roslyn bit directly since it pulls in more dependencies
+                    // than the view engine needs (Microsoft.Extensions.Runtime) for example
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: Is this comment still relevant?

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs (hunk 2 above)
Comment (pakrym) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: Defines are also there

Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: Defines and language version

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs (hunk 3 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: sorta, with renames

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: Are those interesting though? Do we expect people to use `#ifdef` in a view?

Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: Don't see why that matters? What if they do? This basically exposes everything that we use as part of the compilation process.

We at least have a single test that uses it :smile:

**Comment by davidfowl** [comment]
:alien:

**Comment by pakrym** [comment]
Precompilation gets out of sync but i guess we don't care for it much at the moment?

**Comment by davidfowl** [comment]
No we don't care.

**Comment by davidfowl** [comment]
Plus the only way to make it never go out of sync is to only have it be configurable via a file.

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs [hunk 4]
```diff
@@ -22,7 +22,7 @@ public class RazorViewEngineOptionsSetup : ConfigureOptions<RazorViewEngineOptio
         /// <param name="applicationEnvironment"><see cref="IApplicationEnvironment"/> for the application.</param>
         /// <param name="hostingEnvironment"><see cref="IHostingEnvironment"/> for the application.</param>
         public RazorViewEngineOptionsSetup(IApplicationEnvironment applicationEnvironment,
-            IHostingEnvironment hostingEnvironment)
+                                           IHostingEnvironment hostingEnvironment)
```
Comment (pakrym) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs]: nit: default VS whitespace

**Comment by pakrym** [comment]
LGTM

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs [hunk 5]
```diff
@@ -258,13 +254,16 @@ private List<MetadataReference> GetApplicationReferences()
             }
 
             var export = _libraryExporter.GetAllExports(_environment.ApplicationName);
-            foreach (var metadataReference in export.MetadataReferences)
+            if (export != null)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: Curious: Why'd we need to add this?

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/SyntaxTreeGenerator.cs [hunk 6]
```diff
@@ -17,6 +17,19 @@ public static class SyntaxTreeGenerator
             string path,
             CompilationSettings compilationSettings)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/SyntaxTreeGenerator.cs]: Looking at this now, it doesn't really make sense to have a method that doesn't do any compilation take in compilation settings. It'd probably be best removing this overload and just rely on the parseoptions variations.

Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/SyntaxTreeGenerator.cs]: Eventually

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs (hunk 5 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: Because it can be null.

Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/RoslynCompilationService.cs]: It couldn't before?

**Pull request closed without merging**